src/components/tl/mlx5/tl_mlx5_context.c

```diff
@@ -61,8 +69,166 @@ ucc_tl_mlx5_get_context_attr(const ucc_base_context_t *context, /* NOLINT */
+            if (errno != 0 || pos == end_pos) {
+                tl_error(ctx->super.super.lib, "wrong device's port number");
+                return UCC_ERR_INVALID_PARAM;
+            }
+        }
+        strncpy(tmp, ib_devname, devname_len);
+        tmp[devname_len] = '\0';
+        ib_devname       = tmp;
+    }
+    status = ucc_tl_mlx5_create_ibv_ctx(&ib_devname, &ctx->shared_ctx,
+                                        ctx->super.super.lib);
+    if (UCC_OK != status) {
+        tl_error(ctx->super.super.lib, "failed to allocate ibv_context");
+        return UCC_ERR_NO_RESOURCE;
+    }
+    if (port == -1) {
+        port = ucc_tl_mlx5_get_active_port(ctx->shared_ctx);
+    }
+    ctx->ib_port = port;
+    if (-1 == port || !ucc_tl_mlx5_check_port_active(ctx->shared_ctx, port)) {
+        tl_error(ctx->super.super.lib, "no active ports found on %s",
+                 ib_devname);
+        goto destroy_context;
+    }
+    tl_debug(ctx->super.super.lib, "using %s:%d", ib_devname, port);
+
+    ctx->shared_pd = ibv_alloc_pd(ctx->shared_ctx);
+    if (!ctx->shared_pd) {
+        tl_error(ctx->super.super.lib, "failed to allocate ib_pd");
+        goto destroy_context;
+    }
+
     return UCC_OK;
+
+destroy_context:
+    ibv_close_device(ctx->shared_ctx);
+
+    return UCC_ERR_NO_RESOURCE;
+}
+
+ucc_status_t ucc_tl_mlx5_context_create_epilog(ucc_base_context_t *context)
+{
+    ucc_tl_mlx5_context_t *ctx = ucc_derived_of(context, ucc_tl_mlx5_context_t);
+    ucc_context_t *  core_ctx       = context->ucc_context;
+    const char *     template       = "/tmp/ucc.mlx5.XXXXXX";
+    const char *     sockname       = "/sock";
+    size_t           sock_dir_len   = strlen(template) + 1;
+    size_t           sock_path_len  = sock_dir_len + strlen(sockname);
+    int              sock           = 0;
+    char             sock_path[sock_path_len];
+    ucc_subset_t     s;
+    ucc_status_t     status;
+    ucc_topo_t *     topo;
+    ucc_sbgp_t *     sbgp;
+    ucc_tl_team_t *  steam;
+    ucc_coll_task_t *req;
+
+    ucc_assert(core_ctx->service_team != NULL);
+    ucc_assert(core_ctx->params.mask & UCC_CONTEXT_PARAM_FIELD_OOB);
+
+    s.map.type   = UCC_EP_MAP_FULL;
+    s.map.ep_num = core_ctx->params.oob.n_oob_eps;
+    s.myrank     = core_ctx->rank;
+
+    status = ucc_topo_init(s, core_ctx->topo, &topo);
+    if (UCC_OK != status) {
+        tl_error(context->lib, "failed to init mlx5 ctx topo");
+        return status;
+    }
+
+    sbgp = ucc_topo_get_sbgp(topo, UCC_SBGP_NODE);
+    if (sbgp->status != UCC_SBGP_ENABLED) {
+        status = UCC_OK;
+        goto out;
+    }
+
+    ctx->is_imported = sbgp->group_rank != PD_OWNER_RANK;
+
+    if (!ctx->is_imported) {
+        ucc_strncpy_safe(sock_path, template, sock_dir_len);
+        if (mkdtemp(sock_path) != NULL) {
+            status = ucc_tl_mlx5_ib_ctx_pd_init(ctx);
+            if (status != UCC_OK) {
+                tl_error(context->lib, "failed to create ib ctx and pd");
```

Comment:
Ok, so what do you suggest exactly? To remove the `tl_error` ?
About the returned error, in the current implementation we return `UCC_ERR_NO_RESOURCE`, but in your [other comment](https://github.com/openucx/ucc/pull/749#discussion_r1155558296) you suggest returning the actual returned status of `ucc_tl_mlx5_create_ibv_ctx` which may be `UCC_ERR_NOT_FOUND`. What to do about that ?

Reply:
In case we have TL MLX5 compiled and we ran UCC in non IB environment or shared PD is not supported we should fallback to other TLs unless TL MLX5 is forced
any UCC error should work